src/NBitcoin/Block.cs

```diff
@@ -146,18 +177,21 @@ public uint256 GetHash(NetworkOptions options)
             return hash;
         }
 
-        public uint256 GetPoWHash()
+        public virtual uint256 GetPoWHash()
         {
-            return HashX13.Instance.Hash(this.ToBytes());
+            return this.GetHash();
         }
 
         /// <summary>
-        /// If called, <see cref="GetHash"/> becomes cached, only use if you believe the instance will
-        /// not be modified after calculation. Calling it a second type invalidate the cache.
+        /// Precompute the block header hash so that later calls to <see cref="GetHash()"/> will returns the precomputed hash.
         /// </summary>
-        public void CacheHashes()
+        /// <param name="invalidateExisting">If true, the previous precomputed hash is thrown away, else it is reused.</param>
+        /// <param name="lazily">If <c>true</c>, the hash will be calculated and cached at the first call to GetHash(), else it will be immediately.</param>
+        public void PrecomputeHash(bool invalidateExisting = false, bool lazily = false)
         {
-            this.hashes = new uint256[1];
+            this.hashes = invalidateExisting ? new uint256[1] : this.hashes ?? new uint256[1];
```

Comment:
The code may be more readable if replaced with:
```
if (this.hashes == null || invalidateExisting)
   this.hashes = new uint256[1];
```

Reply:
absolutely